rfc/0003-policy-signing.md

```diff
@@ -0,0 +1,440 @@
+
+## Signing Kubewarden policies
+
+We want to sign our policies inside of the GitHub Action that takes care of pushing
+them to our container registry.
+We want to sign them in keyless mode, using the GitHub Action OIDC token.
+
+This is the most secure setup we can imagine.
+
+This leads us to find a solution that solves the problem of the "mutating `Subject`".
+We want end users have a good experience but, at the same time, we don't want to weaken
+the strengths of our signatures.
+
+Given a `Subject`:
+
+```json
+{
+  "Subject": "https://github.com/flavio/policy-secure-pod-images/.github/workflows/release.yml@refs/heads/main"
+}
+```
+
+We can easily understand who is the owner of the repository that built the policy.
+That could be an individual user or an organization.
+
+We plan to allow end users to define an "allow list" of GitHub users/organizations
+that can be trusted.
+
+## Verifying policies
+
+Policy verification settings are going to be defined using a dedicated configuration
+file. This file format is going to be shared between Policy Server and kwctl.
+
+The proposed format is the following one:
+
+```yaml
+public_keys:
+  threshold: 2
+  keys:
+    alice: |
+      -----BEGIN PUBLIC KEY-----
+      MBFKHFDGHKIJH0CAQYIKoZIzj0DAQcDQgAEX0HFTtCfTtPmkx5p1RbDE6HJSGAVD
+      BVDF6SKFSF87AASUspkQsN3FO4iyWodCy5j3o0CdIJD/KJHDJFHDFIu6sA==
+      -----END PUBLIC KEY-----
+    bob: |
+      -----BEGIN PUBLIC KEY-----
+      AUYSDHKJFGSHGSHCAQYIKoZIzj0DAQcDQgAEX0HFTtCfTtPmkx5p1RbtwDE1EVzu
+      wjQs1cCRKb5Pz/yUspkQsN3FO4iyWodCy5j3o0CdIJD/1gvq98pf4IG9tA==
+      -----END PUBLIC KEY-----
+issuers:
+  "https://token.actions.githubusercontent.com":
+  - https://github.com/flavio/
+  - https://github.com/kubewarden/
+  "https://github.com/login/oauth":
+  - john@acme.org
+annotations:
+  env: prod
+  foo: bar
+```
+
+The file is composed by three sections, which we will analyze in the next
+sections.
+
+
+### Public keys
+
+This is a list of the public keys of the authors that are trusted:
+
+```yaml
+public_keys:
+  threshold: 2
+  keys:
+    alice: |
+      -----BEGIN PUBLIC KEY-----
+      MBFKHFDGHKIJH0CAQYIKoZIzj0DAQcDQgAEX0HFTtCfTtPmkx5p1RbDE6HJSGAVD
+      BVDF6SKFSF87AASUspkQsN3FO4iyWodCy5j3o0CdIJD/KJHDJFHDFIu6sA==
+      -----END PUBLIC KEY-----
+    bob: |
+      -----BEGIN PUBLIC KEY-----
+      AUYSDHKJFGSHGSHCAQYIKoZIzj0DAQcDQgAEX0HFTtCfTtPmkx5p1RbtwDE1EVzu
+      wjQs1cCRKb5Pz/yUspkQsN3FO4iyWodCy5j3o0CdIJD/1gvq98pf4IG9tA==
+      -----END PUBLIC KEY-----
+    john:|
+      -----BEGIN PUBLIC KEY-----
+      MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAELKhD7F5OKy77Z582Y6h0u1J3GNA+
+      kvUsh4eKpd1lwkDAzfFDs7yXEExsEkPPuiQJBelDT68n7PDIWB/QEY7mrA==
+      -----END PUBLIC KEY-----
+```
+
+* `keys`: list of keys that are trusted. The owner's name is not relevant during
+  the verification process.
+* `threshold`: an integer number of keys whose signatures are required in order
```

Comment:
I'm not 100% sure about this value, I fear it could make everything more complicated

Reply:
I would prefer adding a root key `required_signatures`, where values can be `allOf` (all info listed here needs to be verified for accepting), or `anyOf` (at least 1 of all info listed here needs to be verified for accepting). Reminiscent of Json schema keywords that are the same.

An other option is to add 2 root keys, `allOf` and `anyOf`. Both or 1 of them can be specified, and both accept `pub_keys`, `keyless`, or `annotations` as children:

```yaml
allOf: # all info listed here needs to be verified for accepting
  pub_keys:
    key-name-irrelevant: |
              -----BEGIN PUBLIC KEY-----
              MBFKHFDGHKIJH0CAQYIKoZIzj0DAQcDQgAEX0HFTtCfTtPmkx5p1RbDE6HJSGAVD
              BVDF6SKFSF87AASUspkQsN3FO4iyWodCy5j3o0CdIJD/KJHDJFHDFIu6sA==
              -----END PUBLIC KEY-----

anyOf: # at least 1 of all info listed here needs to be verified for accepting
  keyless:
    - issuer: https://token.actions.githubusercontent.com # default tuple, we add it to our minimum config
      email: kubewarden@suse.de
  annotations:
    env: prod
    foo: bar
```

But with this second option we need to think if we could create any logic inconsistency. Which I think we do not, unless we add a `oneOf` key, that would signify "only one of the info provided".

Reply:
I like this proposal, but I'm not convinced about adding `annotations` under `allOf` and `anyOf`. I would leave them as top level.

Reply:
So, to summarize:

I would use this format:


```yaml
allOf: # all info listed here needs to be verified for accepting
  pub_keys:
    key-name-irrelevant: |
              -----BEGIN PUBLIC KEY-----
              MBFKHFDGHKIJH0CAQYIKoZIzj0DAQcDQgAEX0HFTtCfTtPmkx5p1RbDE6HJSGAVD
              BVDF6SKFSF87AASUspkQsN3FO4iyWodCy5j3o0CdIJD/KJHDJFHDFIu6sA==
              -----END PUBLIC KEY-----

anyOf: # at least 1 of all info listed here needs to be verified for accepting
  issuers:
    "https://token.actions.githubusercontent.com":
    - url: https://github.com/flavio/
    - url: https://github.com/kubewarden # <- this becomes https://github.com/kubewarden/
    "https://github.com/login/oauth":
    - email: john@acme.org
 
annotations:
 env: prod
 foo: bar
```

Reply:
Closing, this is addressed with the latest iteration of the doc